src/clients/BundleDataClient/utils/DataworkerUtils.ts

```diff
@@ -297,9 +306,62 @@ export function _buildPoolRebalanceRoot(
     });
   });
 
-  // Add to the running balance value from the last valid root bundle proposal for {chainId, l1Token}
-  // combination if found.
-  addLastRunningBalance(latestMainnetBlock, runningBalances, clients.hubPoolClient);
+  // Add to the running balance value from the last valid root bundle proposal. A root bundle proposal which is still in liveness and has not yet been executed counts as a valid proposal.
+  // If the mainnetBundleEndBlock for this pool rebalance root corresponds to the pending root bundle or some root bundle before the pending root bundle, then we can fetch running balances directly from `ExecutedRootBundle` events.
+  if (
+    clients.hubPoolClient.getPendingRootBundle()?.bundleEvaluationBlockNumbers[0] === mainnetBundleEndBlock ||
```

Comment:
The idea here is: if the proposal for which we are trying to build a pool rebalance root is either the pending or a historical proposal, then do what we normally do. Otherwise, this proposal must be "optimistic" so re-calculate the running balance by getting the pool rebalance root of the current pending proposal.

Reply:
I think your other comment makes the check constant-time, so IMO keeping the same function arguments as before seems pretty nice to me.